### PR TITLE
fix: set image_version 9.1 for Rocky on Azure 

### PR DESCRIPTION
**Pull request by rodrigodelmonte**
**What problem does this PR solve?**:

The latest Rocky image version on Azure uses the version `9.1.20230215`
```sh
❯ az vm image list --offer rockylinux-9 --all --output table
Architecture    Offer         Publisher                                           Sku           Urn                                                                                        Version
--------------  ------------  --------------------------------------------------  ------------  -----------------------------------------------------------------------------------------  ------------
x64             rockylinux-9  erockyenterprisesoftwarefoundationinc1653071250513  rockylinux-9  erockyenterprisesoftwarefoundationinc1653071250513:rockylinux-9:rockylinux-9:9.0.0         9.0.0
x64             rockylinux-9  erockyenterprisesoftwarefoundationinc1653071250513  rockylinux-9  erockyenterprisesoftwarefoundationinc1653071250513:rockylinux-9:rockylinux-9:9.1.20230215  9.1.20230215
```

**Which issue(s) does this PR fix?**:

* https://d2iq.atlassian.net/browse/D2IQ-96326

**Special notes for your reviewer**:

 Packer using the latest version:
 ```sh
 ==> azure-arm.kib_image: Getting source image id for the deployment ...
==> azure-arm.kib_image:  -> SourceImageName: '/subscriptions/***/providers/Microsoft.Compute/locations/westus/publishers/erockyenterprisesoftwarefoundationinc1653071250513/ArtifactTypes/vmimage/offers/rockylinux-9/skus/rockylinux-9/versions/latest'
 ```
 
 Cluster is created using Rocky 9.1 instead of Rocky 9.0
 
 ```sh
 NAME                                        STATUS   ROLES           AGE     VERSION   INTERNAL-IP   EXTERNAL-IP   OS-IMAGE                      KERNEL-VERSION                     CONTAINER-RUNTIME
rocky-90-default-b261-control-plane-4kznn   Ready    control-plane   9m16s   v1.25.4   10.0.0.4      <none>        Rocky Linux 9.1 (Blue Onyx)   5.14.0-162.12.1.el9_1.0.2.x86_64   containerd://1.6.17-d2iq.1
rocky-90-default-b261-control-plane-cpqjm   Ready    control-plane   4m18s   v1.25.4   10.0.0.6      <none>        Rocky Linux 9.1 (Blue Onyx)   5.14.0-162.12.1.el9_1.0.2.x86_64   containerd://1.6.17-d2iq.1
rocky-90-default-b261-control-plane-swj7b   Ready    control-plane   6m47s   v1.25.4   10.0.0.5      <none>        Rocky Linux 9.1 (Blue Onyx)   5.14.0-162.12.1.el9_1.0.2.x86_64   containerd://1.6.17-d2iq.1
rocky-90-default-b261-md-0-dc9mm            Ready    <none>          7m12s   v1.25.4   10.1.0.5      <none>        Rocky Linux 9.1 (Blue Onyx)   5.14.0-162.12.1.el9_1.0.2.x86_64   containerd://1.6.17-d2iq.1
rocky-90-default-b261-md-0-k5bwl            Ready    <none>          7m18s   v1.25.4   10.1.0.7      <none>        Rocky Linux 9.1 (Blue Onyx)   5.14.0-162.12.1.el9_1.0.2.x86_64   containerd://1.6.17-d2iq.1
rocky-90-default-b261-md-0-ql78l            Ready    <none>          7m13s   v1.25.4   10.1.0.4      <none>        Rocky Linux 9.1 (Blue Onyx)   5.14.0-162.12.1.el9_1.0.2.x86_64   containerd://1.6.17-d2iq.1
rocky-90-default-b261-md-0-zzbgz            Ready    <none>          7m      v1.25.4   10.1.0.6      <none>        Rocky Linux 9.1 (Blue Onyx)   5.14.0-162.12.1.el9_1.0.2.x86_64   containerd://1.6.17-d2iq.1
```

After the PR change:

```sh
NAME                                        STATUS   ROLES           AGE   VERSION   INTERNAL-IP   EXTERNAL-IP   OS-IMAGE                      KERNEL-VERSION                CONTAINER-RUNTIME
rocky-90-default-3027-control-plane-fdgbj   Ready    control-plane   16m   v1.25.4   10.0.0.5      <none>        Rocky Linux 9.0 (Blue Onyx)   5.14.0-70.17.1.el9_0.x86_64   containerd://1.6.17-d2iq.1
rocky-90-default-3027-control-plane-gzbfd   Ready    control-plane   14m   v1.25.4   10.0.0.6      <none>        Rocky Linux 9.0 (Blue Onyx)   5.14.0-70.17.1.el9_0.x86_64   containerd://1.6.17-d2iq.1
rocky-90-default-3027-control-plane-klhgq   Ready    control-plane   18m   v1.25.4   10.0.0.4      <none>        Rocky Linux 9.0 (Blue Onyx)   5.14.0-70.17.1.el9_0.x86_64   containerd://1.6.17-d2iq.1
rocky-90-default-3027-md-0-9j9sn            Ready    <none>          16m   v1.25.4   10.1.0.4      <none>        Rocky Linux 9.0 (Blue Onyx)   5.14.0-70.17.1.el9_0.x86_64   containerd://1.6.17-d2iq.1
rocky-90-default-3027-md-0-mj8d2            Ready    <none>          16m   v1.25.4   10.1.0.7      <none>        Rocky Linux 9.0 (Blue Onyx)   5.14.0-70.17.1.el9_0.x86_64   containerd://1.6.17-d2iq.1
rocky-90-default-3027-md-0-ppwtn            Ready    <none>          16m   v1.25.4   10.1.0.6      <none>        Rocky Linux 9.0 (Blue Onyx)   5.14.0-70.17.1.el9_0.x86_64   containerd://1.6.17-d2iq.1
rocky-90-default-3027-md-0-qq2f7            Ready    <none>          16m   v1.25.4   10.1.0.5      <none>        Rocky Linux 9.0 (Blue Onyx)   5.14.0-70.17.1.el9_0.x86_64   containerd://1.6.17-d2iq.1
```

**Does this PR introduce a user-facing change?**:

```release-note
It hardcode the image_version instead of use the latest version, which points to a higher version.
```

### CODE REVIEW

**Comment by faiq** [comment]
Lets pin this to 9.1 to be consistent with all other cloud providers

File: images/azure/rocky-90.yaml [hunk 1]
```diff
@@ -6,7 +6,7 @@ packer:
   distribution_version: "rockylinux-9"  # SKU
   # Azure Rocky linux official image: https://portal.azure.com/#view/Microsoft_Azure_Marketplace/GalleryItemDetailsBladeNopdl/id/erockyenterprisesoftwarefoundationinc1653071250513.rockylinux-9
   image_publisher: "erockyenterprisesoftwarefoundationinc1653071250513"
-  image_version: "latest"
+  image_version: "9.0.0"
```
Comment (faiq) [inline images/azure/rocky-90.yaml:9]:
```suggestion
  image_version: "9.1.0"
```

**Comment by dkoshkin** [comment]
Should we kick off the e2e tests?

File: images/azure/rocky-91.yaml [hunk 2]
```diff
@@ -6,7 +6,7 @@ packer:
   distribution_version: "rockylinux-9"  # SKU
   # Azure Rocky linux official image: https://portal.azure.com/#view/Microsoft_Azure_Marketplace/GalleryItemDetailsBladeNopdl/id/erockyenterprisesoftwarefoundationinc1653071250513.rockylinux-9
   image_publisher: "erockyenterprisesoftwarefoundationinc1653071250513"
-  image_version: "latest"
+  image_version: "9.1.0"
```
Comment (supershal) [inline images/azure/rocky-91.yaml:9]: I think this might not work. we will need `9.1.20230215` from the output given in the PR description.

Comment (faiq) [reply images/azure/rocky-91.yaml]: yep

**Comment by faiq** [comment]
running locally and it looks good 

```sh
$ konvoy-image build azure --client-id ${AZURE_CLIENT_ID} --tenant-id ${AZURE_TENANT_ID} images/azure/rocky-91.yaml 
2023/03/13 17:04:59 writing new packer configuration to work/rocky-90-az-1678727099-IRufZ
2023/03/13 17:05:10 starting packer build
rockylinux-9-rockylinux-9: output will be in this color.

==> rockylinux-9-rockylinux-9: Running builder ...
==> rockylinux-9-rockylinux-9: Getting tokens using client secret
==> rockylinux-9-rockylinux-9: Getting tokens using client secret
```

File: ansible/roles/providers/tasks/azure-redhat.yml [hunk 3]
```diff
@@ -15,6 +15,27 @@
...
+    src:  etc/udev/rules.d/66-azure-storage.rules
+    mode: 0644
+  when: ansible_distribution == "Rocky"
+
+- name: unlink /dev/disk/azure
+  command: unlink /dev/disk/azure
+  when: ansible_distribution == "Rocky"
+
+- name: mkdir /dev/disk/azure
+  command: mkdir -p /dev/disk/azure
```
Comment (faiq) [inline ansible/roles/providers/tasks/azure-redhat.yml:30]:
```suggestion
  command: udevadm control --reload-rules && udevadm trigger && mkdir -p /dev/disk/azure
```

File: ansible/roles/providers/templates/etc/udev/rules.d/66-azure-storage.rules [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+ENV{DEVTYPE}=="disk", PROGRAM="/bin/sh -c 'readlink /sys/class/block/%k/device|cut -d: -f4'", ENV{fabric_name}="$env{fabric_scsi_controller}/lun$result"
+
+ENV{fabric_name}=="scsi0/lun0", ENV{fabric_name}="root"
+ENV{fabric_name}=="scsi0/lun1", ENV{fabric_name}="resource"
+# Don't create a symlink for the cd-rom.
+ENV{fabric_name}=="scsi0/lun2", GOTO="walinuxagent_end"
+
+# Create the symlinks.
+LABEL="wa_azure_names"
+ENV{DEVTYPE}=="disk", PROGRAM+="/bin/sh -c '[ -L /dev/disk/azure ] && unlink /dev/disk/azure'", SYMLINK+="disk/azure/$env{fabric_name}"
```
Comment (supershal) [inline ansible/roles/providers/templates/etc/udev/rules.d/66-azure-storage.rules:31]: After a VM is rebooted, the `/dev/disk/azure` is reseted back to `../sda`.
I am unable to figure out which process creates this link. I have ruled out `systemd-udevd`, `cloud-init` and `waagent` services with some testing. 

Following is the hack for working around unlinking the file.
created https://d2iq.atlassian.net/browse/D2IQ-96374 to remove the hack in future.

moving logic to remove symlinks in udev rules `PROGRAM+="/bin/sh -c '[ -L /dev/disk/azure ] && unlink /dev/disk/azure'"` to ensure that rules run upon reboots.

**Comment by supershal** [comment]
Tested manually by creating a VM using azure image.
```
root@shalin-csi-test-1 disk]# tree azure
azure
├── resource -> ../../sdb
├── resource-part1 -> ../../sdb1
├── root -> ../../sda
├── root-part1 -> ../../sda1
├── root-part2 -> ../../sda2
├── root-part3 -> ../../sda3
├── root-part4 -> ../../sda4
└── root-part5 -> ../../sda5
```
More details in jira ticket beacause they cant be provided here.

**Comment by supershal** [comment]
After a VM is rebooted, the /dev/disk/azure is reseted back to ../sda.
I am unable to figure out which process creates this link. I have ruled out systemd-udevd, cloud-init and waagent services with some testing.

Following is the hack for working around unlinking the file.
created https://d2iq.atlassian.net/browse/D2IQ-96374 to remove the hack in future.

moving logic to remove symlinks in udev rules PROGRAM+="/bin/sh -c '[ -L /dev/disk/azure ] && unlink /dev/disk/azure'" to ensure that rules run upon reboots.

File: ansible/roles/providers/tasks/azure-redhat.yml [hunk 5]
```diff
@@ -24,6 +24,15 @@
     mode: 0644
   when: ansible_distribution == "Rocky"
 
+- name: expose cut utility to initramfs setup
```
Comment (supershal) [inline ansible/roles/providers/tasks/azure-redhat.yml:27]: we should move all 4 tasks under `when: ansible_distribution == "Rocky" block.

Comment (supershal) [reply ansible/roles/providers/tasks/azure-redhat.yml:27]: or `ansible_distribution_major_version|int >= 9` this will apply to all redhat

**Comment by supershal** [comment]
Tested the fix on azure cluster using statefulset.
```
Events:
  Type    Reason                  Age   From                     Message
  ----    ------                  ----  ----                     -------
  Normal  Scheduled               110s  default-scheduler        Successfully assigned e2e-volume-pvc-bound-to-pv/volume-pvc-bound-to-pv-tester to rocky-91-udev-shalin-md-0-9gg9k
  Normal  SuccessfulAttachVolume  99s   attachdetach-controller  AttachVolume.Attach succeeded for volume "pvc-f3ada349-6650-4815-af45-4a4f6e1fde63"
  Normal  Pulling                 96s   kubelet                  Pulling image "docker.io/bitnami/kubectl:latest"
  Normal  Pulled                  90s   kubelet                  Successfully pulled image "docker.io/bitnami/kubectl:latest" in 6.194277441s
  Normal  Created                 90s   kubelet                  Created container pvc-haver
  Normal  Started                 90s   kubelet                  Started container pvc-haver

```

**Review by supershal — APPROVED** [review]

File: ansible/roles/providers/tasks/azure-redhat.yml [hunk 6]
```diff
@@ -15,6 +15,20 @@
...
   command: rpm --import https://packages.microsoft.com/keys/microsoft.asc
 
+- block:
+    # TODO: remove cut utility fix to initramfs after its been fixed upstream
+    # https://bugzilla.redhat.com/show_bug.cgi?id=2165042
+    - name: expose cut utility to initramfs setup
+      shell: echo 'install_items+=" /usr/bin/cut "' > /etc/dracut.conf.d/cut.conf
+
+    - name: refresh dracut
+      shell: dracut -f
```
Comment (faiq) [inline ansible/roles/providers/tasks/azure-redhat.yml:25]: implements the solution found here https://access.redhat.com/solutions/6998802

**Pull request merged**